File: fsrs-c/tests/crosscheck.c
/*
 * crosscheck.c — 输出一批状态转移结果（CSV），供 scripts/crosscheck.py 与
 * 独立 Python 参考实现逐项比对。用于验证 C 内核与公式基线无转写偏差。
 *
 * 用法：make crosscheck
 */
#include "fsrs.h"

#include <stdio.h>

static const double kStability[] = { 0.01, 0.1, 1.0, 5.0, 10.0, 100.0, 1000.0, 36500.0 };
static const double kDifficulty[] = { 1.0, 2.5, 5.0, 7.5, 10.0 };
static const uint32_t kDays[] = { 0, 1, 3, 10, 30, 365, 3650 };
static const int kRatings[] = { FSRS_RATING_AGAIN, FSRS_RATING_HARD,
                                FSRS_RATING_GOOD, FSRS_RATING_EASY };

int main(void)
{
    size_t i, j, k, m;
    float desired_retention = 0.9f;

    printf("rating,stability_in,difficulty_in,elapsed_days,"
           "stability_out,difficulty_out,stability_fast_out,interval\n");

    /* 新卡（无历史状态） */
    for (m = 0; m < 4; m++) {
        struct fsrs_next_states ns;
        const struct fsrs_memory_state *st;
        if (fsrs_next_states(NULL, desired_retention, NULL, 0, &ns) != FSRS_OK) {
            return 1;
        }
        st = fsrs_next_states_pick(&ns, kRatings[m]);
        printf("%d,%.17g,%.17g,%u,%.17g,%.17g,%.17g,%u\n",
               kRatings[m], 0.0, 0.0, 0u,
               (double)st->stability, (double)st->difficulty,
               (double)st->stability_fast,
               fsrs_next_states_interval(&ns, kRatings[m]));
    }

    for (i = 0; i < sizeof(kStability) / sizeof(kStability[0]); i++) {
        for (j = 0; j < sizeof(kDifficulty) / sizeof(kDifficulty[0]); j++) {
            for (k = 0; k < sizeof(kDays) / sizeof(kDays[0]); k++) {
                struct fsrs_memory_state cur;
                struct fsrs_next_states ns;
                cur.stability = (float)kStability[i];
                cur.difficulty = (float)kDifficulty[j];
                cur.stability_fast = (float)kStability[i];

                if (fsrs_next_states(NULL, desired_retention, &cur, kDays[k], &ns) != FSRS_OK) {
                    printf("ERROR at S=%g D=%g t=%u\n", kStability[i], kDifficulty[j], kDays[k]);
                    return 1;
                }
                for (m = 0; m < 4; m++) {
                    const struct fsrs_memory_state *st = fsrs_next_states_pick(&ns, kRatings[m]);
                    printf("%d,%.17g,%.17g,%u,%.17g,%.17g,%.17g,%u\n",
                           kRatings[m], kStability[i], kDifficulty[j], kDays[k],
                           (double)st->stability, (double)st->difficulty,
                           (double)st->stability_fast,
                           fsrs_next_states_interval(&ns, kRatings[m]));
                }
            }
        }
    }
    return 0;
}

File: fsrs-c/scripts/crosscheck.py
#!/usr/bin/env python3
"""
把 C 内核的输出与独立 Python 参考实现逐项比对。

用法：
    make crosscheck        # 内部会先跑 C 生成 CSV，再由本脚本比对
    # 或手动：
    ./build/crosscheck > /tmp/c.csv && python3 scripts/crosscheck.py /tmp/c.csv
"""
import csv
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import reference as R  # noqa: E402

REL_TOL = 1e-5   # 相对容差：C 内部用 double 计算但以 float 落盘
ABS_TOL = 1e-6


def rel_err(got, expect):
    denom = max(1.0, abs(expect))
    return abs(got - expect) / denom


def main(path):
    rows = 0
    worst_s = 0.0
    worst_d = 0.0
    interval_mismatch = 0
    failures = []

    with open(path) as fh:
        reader = csv.DictReader(fh)
        for row in reader:
            rating = int(row["rating"])
            s_in = float(row["stability_in"])
            d_in = float(row["difficulty_in"])
            days = int(row["elapsed_days"])
            s_out = float(row["stability_out"])
            d_out = float(row["difficulty_out"])
            sf_out = float(row["stability_fast_out"])
            interval = int(row["interval"])

            state = None if s_in <= 0 else (s_in, d_in, s_in)
            exp_s, exp_d, exp_sf = R.step(state, days, rating)
            exp_interval = R.next_interval(exp_s, 0.9)

            es = rel_err(s_out, exp_s)
            ed = rel_err(d_out, exp_d)
            worst_s = max(worst_s, es)
            worst_d = max(worst_d, ed)

            rows += 1
            if es > REL_TOL and abs(s_out - exp_s) > ABS_TOL:
                failures.append(f"S mismatch rating={rating} S_in={s_in} D_in={d_in} "
                                f"t={days}: C={s_out!r} py={exp_s!r}")
            if ed > REL_TOL and abs(d_out - exp_d) > ABS_TOL:
                failures.append(f"D mismatch rating={rating} S_in={s_in} D_in={d_in} "
                                f"t={days}: C={d_out!r} py={exp_d!r}")
            if rel_err(sf_out, exp_sf) > REL_TOL:
                failures.append(f"S_fast mismatch rating={rating} S_in={s_in} t={days}: "
                                f"C={sf_out!r} py={exp_sf!r}")
            if interval != exp_interval:
                interval_mismatch += 1
                failures.append(f"interval mismatch rating={rating} S_in={s_in} "
                                f"D_in={d_in} t={days}: C={interval} py={exp_interval}")

    print(f"compared {rows} state transitions")
    print(f"max relative error: stability={worst_s:.3e} difficulty={worst_d:.3e}")
    print(f"interval mismatches: {interval_mismatch}")
    if failures:
        print(f"\n{len(failures)} FAILURES (showing first 10):")
        for f in failures[:10]:
            print("  " + f)
        return 1
    print("\nOK: C kernel matches the independent Python reference")
    return 0


if __name__ == "__main__":
    src = sys.argv[1] if len(sys.argv) > 1 else "/tmp/fsrs_crosscheck.csv"
    sys.exit(main(src))
